pkg/trustedresources/verify.go

```diff
@@ -100,18 +70,87 @@ func VerifyPipeline(ctx context.Context, pipelineObj v1beta1.PipelineObject, k8s
 		ObjectMeta: pm,
 		Spec:       pipelineObj.PipelineSpec(),
 	}
-	verifiers, err := getVerifiers(ctx, k8s)
+
+	return verifyResource(ctx, &pipeline, k8s, signature, source, policies)
+}
+
+// verifyResource verifies resource which implements metav1.Object by provided signature and public keys from configmap or policies.
+// It will fetch public key from configmap first, if no keys are found then try to fetch keys from VerificationPolicy
+// For verificationPolicies verifyResource will adopt the following rules to do verification:
+// 1. For each policy, check if the resource url is matching any of the `patterns` in the `resources` list. If matched then this policy will be used for verification.
+// 2. If multiple policies are matched, the resource needs to pass all of them to pass verification.
+// 3. To pass one policy, the resource can pass any public keys in the policy.
+func verifyResource(ctx context.Context, resource metav1.Object, k8s kubernetes.Interface, signature []byte, source string, policies []*v1alpha1.VerificationPolicy) error {
+	verifiers, err := verifier.FromConfigMap(ctx, k8s)
 	if err != nil {
-		return err
+		return fmt.Errorf("failed to get verifiers from configmap: %w", err)
+	}
+	if len(verifiers) != 0 {
+		for _, verifier := range verifiers {
+			if err := verifyInterface(resource, verifier, signature); err == nil {
+				return nil
+			}
+		}
+		return fmt.Errorf("%w: resource %s in namespace %s fails verification", ErrorResourceVerificationFailed, resource.GetName(), resource.GetNamespace())
 	}
 
-	for _, verifier := range verifiers {
-		if err := VerifyInterface(pipeline, verifier, signature); err == nil {
-			return nil
+	if len(policies) == 0 {
+		return errorEmptyVerificationConfig
+	}
+
+	matchedPolicies := []*v1alpha1.VerificationPolicy{}
+	for _, p := range policies {
+		for _, r := range p.Spec.Resources {
+			matching, err := regexp.MatchString(r.Pattern, source)
+			if err != nil {
+				return fmt.Errorf("%v: %w", err, errorRegexMatch)
+			}
+			if matching {
+				matchedPolicies = append(matchedPolicies, p)
+				break
+			}
 		}
 	}
+	if len(matchedPolicies) == 0 {
+		return fmt.Errorf("%w: no matching policies are found for resource: %s against source: %s", errorNoMatchedPolicies, resource.GetName(), source)
+	}
 
-	return fmt.Errorf("Pipeline %s in namespace %s fails verification", pipeline.Name, pipeline.Namespace)
+	for _, p := range policies {
```

Comment:
should this be operating on policies or on matchedpolicies?

Reply:
oh thanks! This is a bug when I formatted the code. And I also missed a tests case here. Will add it!

Reply:
I added one policy with different key and pattern to test this